Add support for using `git` CLI when available

sciwhiz12 opened the pull request:
JGit has some feature holes/defects -- in particular, it lacks support for [git work-trees](https://git-scm.com/docs/git-worktree) and has so [for a long time](https://bugs.eclipse.org/bugs/show_bug.cgi?id=477475).

This PR adds support for using the `git` CLI when it is available, with support for falling back to JGit when not. This means features like work-trees are recognized and work properly.

### TODO

- [x] Move changelog generation to use `GitProvider` as well
	- This does mean more breaking changes to the API surface, on top of what's already been broken in `GradleUtils`, due to removing the use of JGit classes.
- [x] Fix license headers
	- Forge Development LLC has no relation to the new files I've committed
- [ ] Thoroughly test out the changes of this PR
	- Initial tests seem promising in their function
- [ ] Consider switching new classes to Groovy with `@CompileStatic`
- [ ] Consider adding `@CompileStatic` to more places in the project

sciwhiz12 committed:
Add initial support for using git CLI when available

JGit has some feature holes/defects -- in particular, it lacks support
for git work-trees and has so for a long time.

This adds initial support for using the git CLI when it is available,
with support for falling back to JGit when not. This means features
like work-trees are recognized and work properly.

Changelog generation has not yet been moved to using GitProvider, as it
requires more work to detach it from its use of JGit classes.

sciwhiz12 committed:
Ignore copies of newly-added source files

sciwhiz12 committed:
Switch from array to ArrayList

I can't grok proper array copying, so a collection it is!

sciwhiz12 committed:
Fix command in CLI provider's fullBranch

git-symbolic-ref modifies the symbolic ref (first argument) to point to
the second argument. A single argument, for a query, is what we want.

marchermans commented:
This might need a toggle/override.
If the configuration cache is enabled, then we are not allowed to execute commands like git for versioning.

zml2008 commented:
you can run external processes by using ExecOperations - and/or a custom ValueSource on a Provider

marchermans commented:
You can't at the moment because the projects version property does not support it.....

Even jgit currently throws up the error as it turns out. Because it invokes git as well

sciwhiz12 commented:
I'm not fully sure what you mean by what isn't supported by the `Project.version` property?

Although the project version property isn't a full `Property<String>`, it is somewhat flexible because it's actually stored as an `Object` which is `toString`ed on use. I don't remember where I did it, but I recall having using that flexibility to set the project version to a `GitVersion` object, which lazily calculates the version when it's `toString`ed.

lukebemish commented:
> This might need a toggle/override. If the configuration cache is enabled, then we are not allowed to execute commands like git for versioning.

If the configuration cache is enabled, jgit is entirely broken too. In fact, using the git cli is easier to get working with the config cache than jgit - just use a ValueSource or two that takes an ExecOperations and call it a day. No clue whether that's even possible to get working with jgit though, so it might be more sensible to *only* use the git cli.

sciwhiz12 committed:
Merge branch 'main' into git-cli

sciwhiz12 committed:
Move to net.neoforged package and update license headers

sciwhiz12 committed:
Update Gradle to 8.7

IntelliJ IDEA refuses to work with any lower Gradle version, for some
reason.

sciwhiz12 committed:
Reimplement changes to GradleUtils

Remaining unimplemented change is removing GradleUtils.openGit, as that
is currently used by ChangelogUtils.

sciwhiz12 committed:
Migrate git info and version calculator sources to GitProvider

The changelog generator value source will be migrated at a later date,
due to its more extensive uses of JGit.

sciwhiz12 committed:
Replace inline `{@return}` with block `@return`

Inline `{@return}` is not in Java 8, which this project uses.

sciwhiz12 committed:
Migrate changelog generator source to GitProvider

The git CLI provider still needs to implement the required method before
it fully works, but the command needed is supplied in a comment for
future work.

shartte committed:
Merge branch 'main' into fork/sciwhiz12/git-cli

# Conflicts:
#	gradle/wrapper/gradle-wrapper.properties

shartte committed:
Refactor to allow better Git error reporting

shartte committed:
Use StringBuffer if we synchronize anyway

shartte committed:
Implement changelog generation

shartte committed:
Add:
- Fastpath for Changelog generation based on listing all tags and calculating the describe-offset ourselves
- Allow building changelog from origin of repository

sciwhiz12 committed:
Merge branch 'main' into git-cli
